DS-3533 Projections continued

cwilper opened the pull request:
This PR mainly improves the HAL embed behavior so that:

* Subresources are not embedded in responses by default; they are only linked
* If `?projection=full` is requested, subresources are embedded, but only up to 2 levels

It also:

* Standardizes the property names used for multi-valued embeds in JSON responses; the property name for the array now always matches the name of the rel (the last part of the "self" uri of the subresource). Prior to this, there were some inconsistencies with this depending on whether the subresource was embedded (e.g., bundles embedded in a items/uuid request) or requested directly (e.g. items/uuid/bundles)
* Drops the previously-used "list" projection (it doesn't make sense now that the default is to not embed anywhere)
* Simplifies the LinkRest annotation to remove no-longer-used properties.
* Updates tests accordingly

This PR also fixes/addresses: 

* https://jira.lyrasis.org/browse/DS-4248
* https://jira.lyrasis.org/browse/DS-4427

cwilper committed:
DS-3533 Make item subresource embeds optional

cwilper committed:
DS-3533 Make bundle subresource embeds optional

cwilper committed:
DS-3533 Make bitstream subresource embeds optional

cwilper committed:
DS-3533 Require read access to get subresources

cwilper committed:
DS-3533 Consistency: Return NO CONTENT when link repo returns null

cwilper committed:
DS-3533 Rename for consistency among link repos

cwilper committed:
DS-3533 Make collection subresource embeds optional

cwilper committed:
DS-3533 Get optional pageable from one place

cwilper committed:
DS-3533 Consistency: Return NOT FOUND when resource not found for subresource request

cwilper committed:
DS-3533 Consistency: Use link repo for eperson groups

cwilper committed:
DS-3533 Make group subresource embeds optional

cwilper committed:
DS-3533 Fix collection license subresource authorization

cwilper committed:
DS-3533 Consistency: Use link repo to GET bitstream bundle subresource

cwilper committed:
DS-3533 Avoid embedding subresources that may nest (groups in groups, communities in communities)

benbosman commented:
Thanks @cwilper for this PR.

One thing I noticed from your description without having completed a full review:
* You mention renaming "mappedCollections" to "collections". This may become confusing since this doesn't contain the owning collection

Am I correct to summarize this PR as:
* When requesting a single item, the embedded data (relationships, owningCollection, bundles) is excluded when using `projection=list`
* For discovery searches, the embedded data of the items is always excluded
* When requesting item lists, the embedded data of the items is always excluded
* When requesting a single community, the embedded data (collections) is excluded when using `projection=list`
* When requesting community lists, the embedded data of the communities is always excluded

cwilper commented:
Thanks for taking a look @benbosman 

> One thing I noticed from your description without having completed a full review:
> 
> * You mention renaming "mappedCollections" to "collections". This may become confusing since this doesn't contain the owning collection

The "rel" name of the subresource is still mappedCollections -- this just changes the name of the embedded array property inside the "page" of mappedCollections. For this PR, the fact that this inner-array name changed was actually a side-effect of making this info come from a link repository instead of its own controller and related scaffolding. As long as we're consistent, I figured it would be a low friction way to go, as that his how other link repository-provided subresource embeds currently work (using the plural form of the type). But I can also see the argument for forcing it to match the rel name in all cases. I don't mind either style -- I mainly just care about consistency.

> Am I correct to summarize this PR as:
> 
> * When requesting a single item, the embedded data (relationships, owningCollection, bundles) is excluded when using `projection=list`

true.

> * For discovery searches, the embedded data of the items is always excluded

by default yes, because the default projection in that context is "list". But that can be overridden by forcing it via projection=default.

> * When requesting item lists, the embedded data of the items is always excluded

by default, yes, with same exception described above.

> * When requesting a single community, the embedded data (collections) is excluded when using `projection=list`

yes.

> * When requesting community lists, the embedded data of the communities is always excluded

yes.

cwilper commented:
@benbosman I just wanted to quickly follow up on the paged embedded sub-resource rel naming issue and note that I'm looking at consistency issues in detail.

I found an inconsistency with this branch currently: If I temporarily allow certain paged subresources to be embedded in a resource (e.g. mappedCollections embedded in a collection resource), I see that the array name matches the rel name in that context. However, if I hit the URL for the subresource directly, the embedded array name matches the plural type name of the embedded resources. This inconsistency should be addressed. I'm not sure how yet.

I am looking at the possibility of just always making the rel name match the array name regardless of what context the embed occurs in (inside a resource vs. as a subresource via a link repository).

There are other possibilities but I want to explore this one fully before falling back to another approach.

Marking this PR as in-progress again.

benbosman commented:
@cwilper what would this imply for the sub-resources (e.g. item's mappedCollections or browse's entries also named browseEntries)

Currently, an item's mappedCollections is not directly embedded, but a HAL link to /server/api/core/items/uuid/mappedCollections is included. Opening /server/api/core/items/uuid/mappedCollections contains:
```
"_embedded": {
    "mappedCollections": [
...
]
  }
```

A browse's entries is embedded, using a HAL link to e.g. /server/api/discover/browses/subject/entries and the embed of /server/api/discover/browses/subject is:
```
"_embedded": {
    "entries": {
      "_embedded": {
        "entries": []
      },
      "page": {
        "number": 0,
        "size": 20,
        "totalPages": 0,
        "totalElements": 0
      },
      "_links": {
        "self": {
          "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/subject/entries"
        }
      }
    }
  }
```
Opening /server/api/discover/browses/author/entries?size=1 contains:
```
{
  "_embedded": {
    "browseEntries": [
      {
        "authority": null,
        "value": "Abdominal and Paediatric Surgery",
        "valueLang": null,
        "count": 9,
        "type": "browseEntry",
        "_links": {
          "items": {
            "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/author/items?filterValue=Abdominal and Paediatric Surgery"
          }
        }
      }
    ]
  },
  "_links": {
    "first": {
      "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/author/entries?page=0&size=1"
    },
    "self": {
      "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/author/entries?size=1"
    },
    "next": {
      "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/author/entries?page=1&size=1"
    },
    "last": {
      "href": "https://dspace7-demo.atmire.com/server/api/discover/browses/author/entries?page=468&size=1"
    }
  },
  "page": {
    "size": 1,
    "totalElements": 469,
    "totalPages": 469,
    "number": 0
  }
}
```

I would expect the end-result to be:
* The HAL link still contains the name explaining the actual functionality (e.g. mappedCollections)
* The embed in the parent record still contains that same name explaining the actual functionality
* The nested embed in the parent record's embed contains the type of object, e.g.
```
"_embedded": {
   "mappedCollections": {
      "_embedded": {
        "collections": []
      }
   }
}
```
* Opening the HAL link contains the type of object, similar to the inner part of the embed

Does that match your expectations?

Raf-atmire committed:
DS-3533 Disable embeds by default

Raf-atmire committed:
DS-3533 Update javadocs

cwilper committed:
DS-3533 Correct javadocs

cwilper committed:
DS-3533 Drop "list" projection; no longer relevant

cwilper committed:
DS-3533 Include optional links by default

cwilper committed:
DS-3533 Fix test to match new expectation

cwilper committed:
DS-3533 Make community subresource embeds optional

cwilper committed:
DS-3533 Consistency: Use link repo for mapped collections

cwilper committed:
DS-3533 Consistency: Use link repo for mapped items

cwilper committed:
DS-3533 Permit embedding group in eperson

cwilper committed:
DS-3533 Request embeds for relevant ITs

AndrewZachWood committed:
DS-3533 Consistency: Use same json for paged subresources

This ensures that when a subresource is a paged collection,
the json is constructed in the same way, whether it is embedded
in its parent, or requested separately.

cwilper committed:
DS-3533 Drop embed/linkOptional properties from LinkRest

cwilper committed:
DS-3533 Drop linkClass; unused

cwilper committed:
DS-3533 Drop extends; not a link repository

AndrewZachWood committed:
DS-3533 Rest IT fixes

cwilper committed:
DS-3533 Fix ConverterService tests based on new expectations

cwilper committed:
DS-3533 Fix expectation browseEntries -> entries

AndrewZachWood committed:
DS-3533 Fix BrowsesResourceControllerIT test

cwilper committed:
DS-3533 Run test with admin perms since permission is now checked

cwilper committed:
DS-3533 Fix formatting

cwilper committed:
DS-3533 Adjust javadocs

cwilper commented:
Note: The Travis failure above is a timeout, very early in the build, while downloading maven dependencies. Probably a service blip on their end. We have an upcoming commit that will trigger it again soon, but as of yesterday, existing tests *were* passing.

cwilper committed:
DS-3533 Drop unused import

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionMappedItemLinkRepository.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public boolean isEmbeddableRelation(Object data, String name) {
+        return false;
```
For our own records, we may want to add an explanatory comment anytime we disable embedding a particular relationship. This would allow us to document why it was done & warn against (re)enabling it.

cwilper replied:
These were mostly workarounds agains over-embedding things due to cost vs. performance concerns or to avoid infinite loops described by DS-4427. However, given the new embed level-limiting behavior implemented in this branch, most overrides of "isEmbeddableRelation" no longer make sense, so have been removed (including this one).

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunitySubcommunityLinkRepository.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public boolean isEmbeddableRelation(Object data, String name) {
+        return false;
```
Again, explanatory comment on this would be nice.  Though in this case, I'm sure this is about ensuring against loading the entire community "tree" for a single community (and protection against infinite loops)

cwilper replied:
Removed; see previous comment about these.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonGroupLinkRepository.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public boolean isEmbeddableRelation(Object data, String name) {
+        return true;
```
This is the default value, so this override can likely be removed.

cwilper replied:
Good catch, removed.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/GroupGroupLinkRepository.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public boolean isEmbeddableRelation(Object data, String name) {
+        return false;
```
Again, explanatory comment would be nice here.  I'm sure this is to protect against infinite loops though.

cwilper replied:
Removed; see previous comment about these.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemMappedCollectionLinkRepository.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public boolean isEmbeddableRelation(Object data, String name) {
+        return false;
```
Same, add an explanatory comment if possible.

cwilper replied:
Removed; see previous comment about these.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java:
```diff
@@ -127,6 +127,10 @@
         }
     }
 
+    public Pageable getPageable(@Nullable Pageable optionalPageable) {
```
Could you add Javadocs to this new method?

cwilper replied:
Good catch. Added javadoc.

tdonohue reviewed (commented):
@cwilper : Gave this a code review today and overall, I think it looks great.  I have some minor suggestions inline below (mostly adding some comments/JavaDocs).

One observation here is that I was a little surprised at the number of  GET based integration tests that need to use `projection=full`.  My expectation was that this would be a little less prevalent overall.  (That said, I can find quite a few that use the default projection as well.)   Since this is the case, it might be nice to add a few basic tests specific to the `FullProjection` and `DefaultProjection` classes to verify what they are returning.  The other integration tests are definitely exercising parts of each class, but not necessarily doing things like checking that a `DefaultProjection` results in no embedded resources (unless I overlooked a test that checks for that).

I have also tested this PR with a few endpoints and can verify that the two projections (default & full) seem to be working as expected!

AndrewZachWood committed:
DS-3533 Create test for embeds and links against DSO REST ITs

cwilper committed:
DS-3533 Make authority entries subresource embeddable

cwilper committed:
DS-3533 Disable embedding of authority entryValues as a direct subresource

cwilper committed:
DS-3533 Begin consolidating embedded/link tests per dso type

cwilper committed:
DS-3533 Fix incorrect comment noted by Tim

cwilper committed:
DS-4427 Only embed up to one level

AndrewZachWood committed:
DS-3533 Refactor projection embeds ITs to match new style

cwilper committed:
DS-3533 Increase max embed level to 3 and fix tests

cwilper committed:
DS-3533 Fix test issue and imports

cwilper commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BundleBitstreamLinkRepository.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.stereotype.Component;
+
+/**
+ * Link repository for "relationships" subresource of an individual bundle.
```
good catch; fixed

cwilper committed:
DS-4427 Re-enable embedding for link repositories in favor of embed level limiting

cwilper committed:
DS-3533 Add javadoc for new method

cwilper committed:
DS-4427 Update tests based on new embed expectations

benbosman reviewed (commented):
I've tested this PR, and can see e.g.:
* Requesting /server/api/core/communities includes the communities but doesn't embed any data linked to the community
* Requesting /server/api/core/communities?projection=full embeds the communities/subcommunities/collections but doesn't embed communities/subcommunities/collections/logo. It does embed communities/collections/logo but not communities/collections/logo/format
* If the logo doesn't exist, it still embeds `"logo": null` so Angular would be able to cache that there's no logo
* It doesn't change the syntax of the mappedCollections anymore
* Angular doesn't seem to have any issues with these changes (at least for the features I've tested so far)
I will perform a code review next, but so far I don't see any problems in my tests

cwilper committed:
Merge branch 'master' into DS-3533_Projections-Continued

Manually resolved conflicts, including removing CollectionDefaultAccessConditionsLinkRepository
because that subresource is no longer expected to be on collections.

tdonohue reviewed (approved):
👍 This is looking great.  Thanks @cwilper .

I did some testing with both the default projection & "full" projection. Both work as expected. I see a maximum of two levels embedded when using "full" and no embeds when using the default.

The code changes/updates also look good to me. 

While I know Projection work may require more tweaks (e.g. we may want to still consider switching to a max of one-level of embedding, and/or specify embeds as noted in [DS-4433](https://jira.lyrasis.org/browse/DS-4433)), I think this is a good foundation for us to build on. So, I'm in favor of moving this PR forward as-is & improving/enhancing (as needed) in later PRs.

benbosman commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorityEntryLinkRepository.java:
```diff
@@ -50,12 +51,12 @@
     @PreAuthorize("hasAuthority('AUTHENTICATED')")
-    public Page<AuthorityEntryRest> query(HttpServletRequest request, String name,
-                                          Pageable pageable, Projection projection) {
+    public Page<AuthorityEntryRest> query(@Nullable HttpServletRequest request, String name,
```
What's the use case we need AuthorityEntryLinkRepository to have a nullable request and pageable?
This change was not clear to me

cwilper replied:
The reason I added these indicators was mainly because there's an implicit contract in link repository getter methods, and I wanted the code to be as explicit (as possible) about that. In general, any time a link repository is called while embedding a subresource into another resource, those can be null.

I tried to resist refactoring too much in these early PRs, but I would like to eventually see some refactoring with link repository methods in general. I think they could adhere to a real java interface (maybe one for paged subresources, and one for scalar ones), in which case, if it was really important to pass down something like the HttpServletRequest, the fact that it could be nullable would be part of the java interface, not necessarily the impl.

benbosman reviewed (approved):
From my code review, I've noticed many of the changes are moving code and refactoring methods. I don't see any problems with those changes

I've verified that the query /server/api/core/communities without projections doesn't request subcommunities or collections from the database, so there's no data loaded which is not required.
Similarly, no logos, … are requested from the database

I've also verified that the query /server/api/core/communities?projection=full (with projections) does request subcommunities, collections, logos, … from the database, but didn't request those when they were not going to be embedded. I did notice for logos (bitstreams) they do request bitstream formats from the DB at all time because they're eager, and do expand some LAZY requests despite being irrelevant, but I added more details about it inline, and verified it was not due to a problem in this PR.

Apart from the small inline questions, this looks great to me. I'm approving them since they're not bugs or problems

cwilper commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java:
```diff
@@ -93,6 +92,11 @@
+    /**
+     * The maximum number of embed levels to allow.
+     */
+    private static final int EMBED_MAX_LEVELS = 2;
```
Yes, it occurred to me that this could be a server-wide setting or even a per-request arg. If there are use cases for that, it shouldn't be hard to add. We should still probably enforce a maximum, though.

tdonohue replied:
I had the same thought on configurability, though we may want to put a lot of warnings around that or enforce a maximum.  In any case though, I honestly feel that could come in a future PR, rather than trying to "perfect" everything in one PR.

mwoodiupui commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java:
```diff
@@ -600,20 +601,29 @@ private void embedMethodLevelRel(HALResource<? extends RestAddressableModel> res
-            return converter.toResource((RestAddressableModel) linkedObject);
+            RestAddressableModel restObject = (RestAddressableModel) linkedObject;
+            restObject.setEmbedLevel(childEmbedLevel);
+            return converter.toResource(restObject);
```
Jira ticket for the latter, so we don't lose track of it?

cwilper replied:
Interesting. I also can't say why a query would have been run for the bundles and communities of the logo bitstreams: The code that limits the level (in Utils.embedRelFromRepository) definitely returns *before* the link repository is called. And Bitstream.java specifies FetchType.LAZY for both relations.

This seems worth following up on at some point to identify the exact cause, as it may be systemic.

cwilper commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java:
```diff
@@ -127,6 +132,16 @@
+     * @return the existing instance if it is not null, a default pageable instance otherwise.
+     */
+    public Pageable getPageable(@Nullable Pageable optionalPageable) {
+        return optionalPageable != null ? optionalPageable : new PageRequest(0, 20);
```
I'd like to see a constant here, too. 

The only reason I didn't use EMBEDDED_PAGE_SIZE was that this code is called in contexts where the page is embedded in a parent resource, and when the page is requested standalone, as a subresource.

However, I think in reality, it's 20 in either case, so maybe we just call it DEFAULT_PAGE_SIZE (whether it's embedded or not)

cwilper committed:
DS-3533 Address PR feedback

cwilper commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -793,7 +794,12 @@ public ResourceSupport findRel(HttpServletRequest request, HttpServletResponse r
-                            .invoke(linkRepository, request, uuid, page, utils.obtainProjection(true));
+                            .invoke(linkRepository, request, uuid, page, utils.obtainProjection());
+
+                    if (pageResult == null) {
```
Glad you spotted this. Actually there is never a good reason for a link repository to return null for a paged subresource (and none do anymore, afaik), so I changed it to log and respond with an internal server error, with comments inline as to why.

tdonohue commented:
Merging, as this is at +2.  Thanks for the (continued) hard work on Projections @cwilper !
